File: buffer-calculation/c-lang/incorrect-buffer-size.c
#include <stdio.h>
#include <string.h>

void vulnerableFunction(char *src) {
    // Incorrectly calculating buffer size
    int bufSize = strlen(src) - 1;
    char dst[bufSize];
    strncpy(dst, src, strlen(src));  // This line could lead to a buffer overflow

    printf("%s\n", dst);
}

int main() {
    char data[] = "hello, world";
    vulnerableFunction(data);
    return 0;
}

File: buffer-calculation/go-lang/incorrect-buffer-size.go
package main

import (
	"fmt"
)

func vulnerableFunction(src []byte) {
	// Incorrectly calculating buffer size
	bufSize := len(src) - 1
	dst := make([]byte, bufSize)
	copy(dst, src)  // This line could lead to an out-of-bounds read if src has a length of 0

	fmt.Println(string(dst))
}

func main() {
	data := []byte("hello, world")
	vulnerableFunction(data)
}
